[Merged by Bors] - feat: miscellaneous Lie algebra lemmas

ocfnash opened the pull request:
These came up in the proof that the Geck Construction yields _semisimple_ Lie algebras. I am splitting them into their own PR to simplify review.


---

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ocfnash committed:
feat: miscellaneous Lie algebra lemmas

These came up in the proof that the Geck Construction yields _semisimple_ Lie
algebras. I am splitting them into their own PR to simplify review.

ocfnash committed:
Appease shake

eric-wieser commented on Mathlib/Algebra/Lie/Basic.lean:
```diff
@@ -296,18 +296,22 @@ def LieRing.toNonUnitalNonAssocRing : NonUnitalNonAssocRing L :=
+
+theorem sum_lie_sum {κ : Type*} (s : Finset ι) (t : Finset κ) (f : ι → L) (g : κ → M) :
+    ⁅(∑ i ∈ s, f i), ∑ j ∈ t, g j⁆ = ∑ i ∈ s, ∑ j ∈ t, ⁅f i, g j⁆ := by
+  simp_rw [lie_sum, sum_lie]
```
Does rewriting in the other order remove the need to commute the sums?

ocfnash replied:
Good point; done.

ocfnash commented on Mathlib/Algebra/Lie/Basic.lean:
```diff
@@ -296,18 +296,22 @@ def LieRing.toNonUnitalNonAssocRing : NonUnitalNonAssocRing L :=
-    ⁅(∑ i ∈ s, f i), ∑ j ∈ t, g j⁆ = ∑ i ∈ s, ∑ j ∈ t, ⁅f i, g j⁆ :=
-  let _i := LieRing.toNonUnitalNonAssocRing L
-  s.sum_mul_sum t f g
+theorem sum_lie (s : Finset ι) (f : ι → L) (m : M) : ⁅∑ i ∈ s, f i, m⁆ = ∑ i ∈ s, ⁅f i, m⁆ := by
```
OK sure; done.

ocfnash committed:
Apply suggestions from CR

ocfnash committed:
Appease shake

eric-wieser commented on Mathlib/Algebra/Lie/Basic.lean:
```diff
@@ -296,18 +295,17 @@ def LieRing.toNonUnitalNonAssocRing : NonUnitalNonAssocRing L :=
-  let _i := LieRing.toNonUnitalNonAssocRing L
-  s.sum_mul f a
+theorem sum_lie (s : Finset ι) (f : ι → L) (m : M) : ⁅∑ i ∈ s, f i, m⁆ = ∑ i ∈ s, ⁅f i, m⁆ := by
+  let g : L →+ M := ⟨⟨fun x ↦ ⁅x, m⁆, zero_lie m⟩, fun x y ↦ add_lie x y m⟩
```
My claim is that this is useful enough to be a global definition!

ocfnash committed:
Introduce `LieModule.toEndAddHom`

The lack of `AddHom.flip` (for interchanging the two arguments of a biadditive
map) means that we cannot conveniently use this to golf `sum_lie`.

eric-wieser commented on Mathlib/Algebra/Lie/Basic.lean:
```diff
@@ -296,18 +304,16 @@ def LieRing.toNonUnitalNonAssocRing : NonUnitalNonAssocRing L :=
-  s.sum_mul f a
+theorem sum_lie (s : Finset ι) (f : ι → L) (m : M) : ⁅∑ i ∈ s, f i, m⁆ = ∑ i ∈ s, ⁅f i, m⁆ :=
+  let g : L →+ M := ⟨⟨fun x ↦ ⁅x, m⁆, zero_lie m⟩, fun x y ↦ add_lie x y m⟩
+  map_sum g f s
```
```suggestion
  map_sum ((LieModule.toEndAddHom L M).flip m) f s
```
probably

ocfnash replied:
Oh my previous commit message was a lie: we have `MonoidHom.flip`.

ocfnash committed:
Golf

ocfnash committed:
Rename: this should not be in the `LieModule` namespace

ocfnash committed:
Merge branch 'master' into ocfnash/lie_lemmas

ocfnash committed:
Fix after merging master

ocfnash committed:
Merge branch 'master' into ocfnash/lie_lemmas

eric-wieser commented on Mathlib/Algebra/Lie/Semisimple/Defs.lean:
```diff
@@ -29,6 +29,14 @@ variable [CommRing R] [LieRing L] [AddCommGroup M] [Module R M] [LieRingModule L
   IsSimpleOrder (LieSubmodule R L M)
 
+variable {R L M} in
+lemma LieModule.IsIrreducible.mk [Nontrivial M] (h : ∀ N : LieSubmodule R L M, N ≠ ⊥ → N = ⊤) :
```
Should this be a constructor for `IsSimpleOrder` instead?

eric-wieser replied:
(or maybe just removed, since this is the existing constructor modulo `or_iff_not_imp_left`)

bors d+

Thanks!

ocfnash committed:
Update Mathlib/Algebra/Lie/Semisimple/Defs.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ocfnash committed:
Convenience constructor for `IsSimpleOrder`

ocfnash committed:
Rename `IsSimpleOrder.mk'` to `IsSimpleOrder.of_forall_eq_top`

ocfnash commented on Mathlib/Order/Atoms.lean:
```diff
@@ -708,6 +708,11 @@ class IsSimpleOrder (α : Type*) [LE α] [BoundedOrder α] : Prop extends Nontri
 
 export IsSimpleOrder (eq_bot_or_eq_top)
 
+lemma IsSimpleOrder.mk' {α : Type*} [LE α] [BoundedOrder α] [Nontrivial α]
```
Done (though I'm dubious about having `of_forall_eq_bot` since I claim `of_forall_eq_top` mostly exists for psychological reasons which are unlikely to apply to `of_forall_eq_bot`).

ocfnash commented:
bors merge

ocfnash committed:
Merge branch 'master' into ocfnash/lie_lemmas

Maybe this will circumvent the CI issue?

ocfnash commented:
Urgh, bors crashed _again_ *and* `lake build --no-build` in CI incorrectly failed 🫠 

bors merge

bryangingechen commented:
Maybe it will work with an empty queue?
bors r+
